src/Sulu/Component/Rest/ListBuilder/Metadata/Doctrine/Driver/XmlDriver.php

```diff
@@ -0,0 +1,256 @@
+    {
+        $type = new ConcatenationType(XmlUtil::getValueFromXPath('@orm:glue', $xpath, $propertyNode, ' '));
+        foreach ($xpath->query('orm:field', $propertyNode) as $fieldNode) {
+            if (($field = $this->getField($xpath, $fieldNode)) === null) {
+                continue;
+            }
+
+            $type->addField($field);
+        }
+
+        return $type;
+    }
+
+    /**
+     * Extracts data from dom-node to create a new field object.
+     *
+     * @param \DOMXPath $xpath
+     * @param \DOMElement $fieldNode
+     *
+     * @return FieldMetadata
+     */
+    protected function getField(\DOMXPath $xpath, \DOMElement $fieldNode)
+    {
+        if (($reference = XmlUtil::getValueFromXPath('@property-ref', $xpath, $fieldNode)) !== null) {
+            $nodeList = $xpath->query(sprintf('/x:class/x:properties/x:*[@name="%s"]', $reference));
+
+            if ($nodeList->length === 0) {
+                return;
+            }
+
+            return $this->getField($xpath, $nodeList->item(0));
+        }
+
+        if (($fieldName = XmlUtil::getValueFromXPath('orm:field-name', $xpath, $fieldNode)) === null
+            || ($entityName = XmlUtil::getValueFromXPath('orm:entity-name', $xpath, $fieldNode)) === null
+        ) {
+            return;
+        }
+
+        $field = new FieldMetadata($this->resolveParameter($fieldName), $this->resolveParameter($entityName));
+
+        $joinsNodeList = $xpath->query('orm:joins', $fieldNode);
+        if ($joinsNodeList->length > 0) {
+            $this->getJoinsMetadata($xpath, $joinsNodeList->item(0), $field);
+        }
+
+        return $field;
+    }
+
+    /**
+     * Extracts data from dom-node to create all join-metadata.
+     *
+     * @param \DOMXPath $xpath
+     * @param \DOMElement $joinsNode
+     * @param FieldMetadata $field
+     */
+    protected function getJoinsMetadata(\DOMXPath $xpath, \DOMElement $joinsNode, FieldMetadata $field)
+    {
+        if (($reference = XmlUtil::getValueFromXPath('@ref', $xpath, $joinsNode)) !== null) {
+            $nodeList = $xpath->query(sprintf('/x:class/orm:joins[@name="%s"]', $reference));
+
+            if ($nodeList->length === 0) {
+                return;
+            }
+
+            $this->getJoinsMetadata($xpath, $nodeList->item(0), $field);
+        }
+
+        foreach ($xpath->query('orm:join', $joinsNode) as $joinNode) {
+            $field->addJoin($this->getJoinMetadata($xpath, $joinNode));
+        }
+    }
+
+    /**
+     * Extracts data from dom-node to create a new join-metadata object.
+     *
+     * @param \DOMXPath $xpath
+     * @param \DOMElement $joinNode
+     *
+     * @return JoinMetadata
+     */
+    protected function getJoinMetadata(\DOMXPath $xpath, \DOMElement $joinNode)
+    {
+        $joinMetadata = new JoinMetadata();
+
+        if (($fieldName = XmlUtil::getValueFromXPath('orm:field-name', $xpath, $joinNode)) !== null) {
```

Comment:
Not important, but [this great library](https://github.com/phpbench/dom) makes working with the DOM and Xpath easier.